examples/amm/src/contract.rs

```diff
@@ -0,0 +1,112 @@
+use amm::{get_pool_balance, update_pool_balance, AMMError};
+use async_trait::async_trait;
+use linera_sdk::{
+    base::{SessionId, WithContractAbi},
+    ApplicationCallResult, CalleeContext, Contract, ExecutionResult, MessageContext,
+    OperationContext, SessionCallResult, ViewStateStorage,
+};
+
+linera_sdk::contract!(Amm);
+
+impl WithContractAbi for Amm {
+    type Abi = amm::AMMAbi;
+}
+
+#[async_trait]
+impl Contract for Amm {
+    type Error = AMMError;
+    type Storage = ViewStateStorage<Self>;
+
+    async fn initialize(
+        &mut self,
+        _context: &OperationContext,
+        _argument: (),
+    ) -> Result<ExecutionResult<Self::Message>, AMMError> {
+        Ok(ExecutionResult::default())
+    }
+
+    async fn execute_operation(
+        &mut self,
+        _context: &OperationContext,
+        operation: Self::Operation,
+    ) -> Result<ExecutionResult<Self::Message>, AMMError> {
+        match operation {
+            Self::Operation::Swap {
+                input_token,
+                output_token,
+                amount,
+            } => {
+                let input_pool_balance = get_pool_balance(&self.token_pool, &input_token).await?;
+                let output_pool_balance = get_pool_balance(&self.token_pool, &output_token).await?;
+
+                let output_amount =
+                    amm::calculate_output_amount(amount, input_pool_balance, output_pool_balance)?;
+
+                update_pool_balance(
+                    &mut self.token_pool,
+                    &input_token,
+                    input_pool_balance - amount,
+                )?;
+                update_pool_balance(
+                    &mut self.token_pool,
+                    &output_token,
+                    output_pool_balance + output_amount,
+                )?;
+
+                Ok(ExecutionResult::default())
+            }
+            Self::Operation::AddLiquidity { token, amount } => {
+                let existing_balance = get_pool_balance(&self.token_pool, &token).await?;
+                let new_balance = existing_balance + amount;
+                update_pool_balance(&mut self.token_pool, &token, new_balance)?;
+
+                Ok(ExecutionResult::default())
+            }
+            Self::Operation::RemoveLiquidity { token, amount } => {
+                let existing_balance = get_pool_balance(&self.token_pool, &token).await?;
+                if amount > existing_balance {
+                    return Err(AMMError::InsufficientLiquidityError);
+                }
+
+                let new_balance = existing_balance - amount;
+                update_pool_balance(&mut self.token_pool, &token, new_balance)?;
+
+                Ok(ExecutionResult::default())
+            }
+        }
+    }
+
+    async fn execute_message(
+        &mut self,
+        _context: &MessageContext,
+        _message: (),
+    ) -> Result<ExecutionResult<Self::Message>, AMMError> {
+        Err(AMMError::MessagesNotSupported)
```

Comment:
I guess we'll replace the strings with `ApplicationId`s in the future (referring to token applications), and then we'll have messages to add and remove actual tokens to/from the liquidity pools? Alternatively (or in addition) the tokens could be required to already be on the same chain, so they could be added or removed with operations and cross-application calls. But for placing bids, we'd still use cross-chain messages.

Reply:
I'll try to fix the existing version to work according to the comments, and then we can talk about adding the messaging and stuff to it (just so I have some incremental building blocks to make sense of it in my head)